### ISSUE TITLE
Explanation of with_shopify_session

### ISSUE THREAD

**Issue by resistorsoftware** [Shopify/shopify_app#907]
I am looking at logs. I have a background job that runs and it stores a value on the shop record, relevant to the shop. 

I open a shop.with_shopify_session block so I can work on the store, and I run the following code:
    
     shop.my_special_atttribute = value
     shop.save

In the logs, for each update call to the DB, the update is preceded by an Exists? function issued by the code in the Shopify App gem, pertaining to the session. So it looks like every single touch to the AR model of Shop in fact issues two DB calls. 

I find this curious as a pattern. Naively I figured if I call with_shopify_session method based off of a shop instance, it would ensure I have a valid API session and away I go. Instead I see it is actually touching the DB on every call involving shop. 

I know I could roll my own code to open a session once, and then do all my operations and complete my work, but I saw this Shopify code as something battle-tested, tried and true, and I figured it must be efficient too. Seeing two calls for one seems less than efficient. 

TL:DR; is it expected that we not use that construct and instead open our own sessions to get things done with less DB traffic? If yes, fine, then why does this construct exist?

**Comment by tanema** [comment]
`with_shopify_session` is only to setup our ActiveResource connection to shopify. It should have nothing to do with your DB. 

Shopify app with shopify session
```ruby
    def with_shopify_session(&block)
      ShopifyAPI::Session.temp(
        domain: shopify_domain,
        token: shopify_token,
        api_version: api_version,
        &block
      )
    end
```

And all it is essentially calling down the stack in ShopifyAPI::Base is 
```ruby
      def activate_session(session)
        raise InvalidSessionError.new("Session cannot be nil") if session.nil?
        self.site = session.site
        self.headers.merge!('X-Shopify-Access-Token' => session.token)
        self.api_version = session.api_version
      end
```

What I imagine is that something is going on with your model validations. Do you see this query before a manual call to update locally? Do you see the query happen at the beginning of the session block or just before calling `save`

**Comment by resistorsoftware** [comment]
I am not validating anything in my model. You can see the code I provided. I do a shop.save which triggers the correct update, but it is always preceded by that Exists? Which seems to be a Shopify thing, perhaps from this concern? I am just curious. I see the logs, and it is on every single update inside the block shop.with_shopify_session.

    module ShopifyApp
        module SessionStorage
          extend ActiveSupport::Concern
              included do
                 validates :shopify_token, presence: true
                 validates :api_version, presence: true
                 validates :shopify_domain, presence: true,
              end

**Comment by tanema** [comment]
Okay yes. I just played around with it and it is the uniqueness validation that runs it here: https://github.com/Shopify/shopify_app/blob/master/lib/shopify_app/session/session_storage.rb#L16

**Comment by resistorsoftware** [comment]
So do you think it is apt to re-factor the concern so that the SELECT under discussion is generated and executed just once when the method call is made, and not on every touch of the shop model? Seeing logs fill with these calls seems to indicate it might be worth it:

    Shop Exists? (2.8ms)  SELECT 1 AS one FROM "shops" WHERE LOWER("shops"."shopify_domain") = LOWER($1) AND "shops"."id" != $2 LIMIT $3  [["shopify_domain", "beer-mcbeer-beer.myshopify.com"], ["id", 1], ["LIMIT", 1]]

**Comment by tanema** [comment]
It is hard to tell which is better. We could add complexity by running the validation only if the domain has changed but exist queries are very low cost and they only run during the save transaction. I can ask the rails maintainers if they recommend something better.

**Comment by resistorsoftware** [comment]
I do appreciate the fact you dug into this for me! It truly is not a stopper of anything in any sense, and as you point out, some minor refactoring could eliminate it altogether, including if I just did one final save instead of a save during every iteration.

**Comment by tanema** [comment]
I am looking to merge this PR https://github.com/Shopify/shopify_app/pull/887 this week which will separate the validation concerns and this validation will be removed, so after this change you should no longer see this behaviour

**Comment by resistorsoftware** [comment]
That is nice! Moving forward with that looks to be a lot more organized!

**Comment by tanema** [comment]
I believe this should be no longer an issue.
